<?xml version="1.0" encoding="utf-8"?>
<launch>
    <!-- Required arguments -->
    <arg name="controllers_ns"      default="anymal_d"/>
    <arg name="controllers_args"    default="joint_controller"/>
    <arg name="tbai_config_path"    default="$(find tbai_ros_deploy_go2_mpc)/config/deploy_go2_mpc.yaml"/>
    <arg name="description_name"    default="robot_description"/>
    <arg name="description_file"    default="$(find ocs2_robotic_assets)/resources/go2/urdf/go2_description.urdf"/>

    <arg name="sqp_settings_file" default="$(find tbai_ros_deploy_go2_mpc)/config/sqp.info"/>
    <arg name="frame_declaration_file" default="$(find tbai_ros_deploy_go2_mpc)/config/frame_declarations.info"/>
    <arg name="task_settings_file" default="$(find tbai_ros_deploy_go2_mpc)/config/task.info"/>
    <arg name="target_command_config_file" default="$(find tbai_ros_deploy_go2_mpc)/config/targetCommand.info"/>
    <arg name="controller_config_file" default="$(find tbai_ros_deploy_go2_mpc)/config/controllers.info"/>
    <arg name="gait_command_file" default="$(find tbai_ros_deploy_go2_mpc)/config/gait.info"/>

    <arg name="network_interface"           default="eth0"/>
    <arg name="run_rviz"                    default="false"/>
    <arg name="run_virtual_joystick"        default="false"/>
    <arg name="run_go2_joystick"            default="true"/>
    <arg name="publish_pointcloud"          default="false"/>
    <arg name="mapping_device"              default="none"/>

    <param name="sqp_settings_file" type="string" value="$(arg sqp_settings_file)"/>
    <param name="frame_declaration_file" type="string" value="$(arg frame_declaration_file)"/>
    <param name="task_settings_file" type="string" value="$(arg task_settings_file)"/>
    <param name="target_command_config_file" type="string" value="$(arg target_command_config_file)"/>
    <param name="controller_config_file" type="string" value="$(arg controller_config_file)"/>
    <param name="gait_command_file" type="string" value="$(arg gait_command_file)"/>

    <!-- Load gazebo description -->
	<param name="$(arg description_name)" command="$(find xacro)/xacro --inorder $(arg description_file)" />

    <!-- Upload config path to ROS param server -->
    <param name="tbai_config_path" type="string" value="$(arg tbai_config_path)"/>

    <!-- Set environment variable for config path -->
    <env name="TBAI_GLOBAL_CONFIG_PATH" value="$(arg tbai_config_path)"/>

    <!-- Set environment variable for robot description path -->
    <env name="TBAI_ROBOT_DESCRIPTION_PATH" value="$(arg description_file)"/>
    
    <!-- Set environment variable for network interface -->
    <env name="TBAI_GO2_NETWORK_INTERFACE" value="$(arg network_interface)"/>

    <env name="TBAI_GO2_PUBLISH_LIDAR" value="$(arg publish_pointcloud)"/>

    <!-- Launch simple node -->
    <node name="deploy_go2_mpc" pkg="tbai_ros_deploy_go2_mpc" type="deploy_go2_mpc" output="screen"/>

    <!-- Launch MPC node -->
    <node pkg="tbai_ros_mpc" type="anymal_mpc_node" name="anymal_mpc_node" output="screen"/>

    <!-- Launch Gait command node -->
    <node pkg="tbai_ros_deploy_go2_mpc" type="go2_gait_node" name="go2_gait_node"
          output="screen" launch-prefix="gnome-terminal --"/>

    <!-- Launch virtual joystick -->
    <group if="$(arg run_virtual_joystick)">
        <node name="virtual_joystick" pkg="tbai_ros_utils" type="virtual_joystick.py" output="screen"/>
    </group>

    <group if="$(arg run_go2_joystick)">
        <node name="go2_joystick_node" pkg="tbai_ros_deploy_go2_mpc" type="go2_joystick_node" output="screen"/>
    </group>

    <!-- Start Rviz visualizer -->
    <group if="$(arg run_rviz)">
        <arg name="rvizconfig" default="$(find tbai_ros_deploy_go2_mpc)/launch/rviz/deploy_go2_mpc.rviz" />
        <node pkg="rviz" type="rviz" name="rviz" args="-d $(arg rvizconfig)" output="screen" />
    </group>

</launch>
